src/xdr_mem.c

```diff
@@ -183,6 +183,8 @@ xdrmem_putbufs(XDR *xdrs, xdr_uio *uio, u_int flags)
 	for (ix = 0; ix < uio->uio_count; ++ix) {
 		xdr_vio *v = &(uio->uio_vio[ix]);
 
+		uio->uio_references++;
```

Comment:
This doesn't look right, this is in a loop...

Also would help to have more explanation as to why we are replacing a decrement of refcount with an increment (in a loop no less). Where is the matching decrement(s)?

Reply:
looks like we are missing to take uio_reference in xdrmem_putbufs

uio_references is 0 when we allocated uio in xdr_READ3res_uio_setup
```
(gdb) 
825       uio = gsh_calloc(1, sizeof(struct xdr_uio) + sizeof(struct xdr_vio));
(gdb) 
826       uio->uio_release = xdr_READ3res_uio_release;
(gdb) 
827       uio->uio_count = 1;
(gdb) 
828       uio->uio_vio[0].vio_base = objp->data.data_val;
(gdb) 
829       uio->uio_vio[0].vio_head = objp->data.data_val;
(gdb) 
830       uio->uio_vio[0].vio_tail = objp->data.data_val + size2;
(gdb) 
831       uio->uio_vio[0].vio_wrap = objp->data.data_val + size2;
(gdb) 
832       uio->uio_vio[0].vio_length = size2;
(gdb) p *uio
$1 = {uio_refer = 0x0, uio_release = 0x7f47c698261d <xdr_READ3res_uio_release>, uio_p1 = 0x0, uio_p2 = 0x0, uio_u1 = 0x0, uio_u2 = 0x0, uio_count = 1, 
  uio_flags = 0, uio_references = 0, uio_vio = 0x7f47ba48f0c0}
(gdb) p uio
$2 = (struct xdr_uio *) 0x7f47ba48f080
```

Now in xdrmem_putbufs we decrement and check uio_references == 0 to release uio, which fails since we never took the ref and after decrementing uio_references = -1
```
Breakpoint 2, xdrmem_putbufs (xdrs=0x7f47ba4191c8, uio=0x7f47ba48f080, flags=0) at /home/nutanix/nfs-ganesha/src/libntirpc/src/xdr_mem.c:177
177       xdr_tail_update(xdrs);
(gdb)
179       __warnx(TIRPC_DEBUG_FLAG_XDR,
(gdb)
183       for (ix = 0; ix < uio->uio_count; ++ix) {
(gdb) p uio
$3 = (xdr_uio *) 0x7f47ba48f080
(gdb) p *uio
$4 = {uio_refer = 0x0, uio_release = 0x7f47c698261d <xdr_READ3res_uio_release>, uio_p1 = 0x0, uio_p2 = 0x0, uio_u1 = 0x0, uio_u2 = 0x0, uio_count = 1,
  uio_flags = 0, uio_references = 0, uio_vio = 0x7f47ba48f0c0}
(gdb) n
184        xdr_vio *v = &(uio->uio_vio[ix]);
(gdb)
186        if (!XDR_PUTBYTES(xdrs, v->vio_head, v->vio_length))
(gdb)
189        __warnx(TIRPC_DEBUG_FLAG_XDR,
(gdb)
183       for (ix = 0; ix < uio->uio_count; ++ix) {
(gdb)
196       if (!(--uio->uio_references)) {
(gdb)
207       return (TRUE);
(gdb) p uio
$5 = (xdr_uio *) 0x7f47ba48f080
(gdb) p *uio
$6 = {uio_refer = 0x0, uio_release = 0x7f47c698261d <xdr_READ3res_uio_release>, uio_p1 = 0x0, uio_p2 = 0x0, uio_u1 = 0x0, uio_u2 = 0x0, uio_count = 1,
  uio_flags = 0, uio_references = -1, uio_vio = 0x7f47ba48f0c0}
(gdb) 
```

So I just added the fix to take ref by incrementing uio_references, we always keep uio_count=1, so uio_references should be 1, xdr_READ3res_uio_release will decrement the ref taken.

```
        if (!(--uio->uio_references)) {
                for (ix = 0; ix < uio->uio_count; ix++) {
                        gsh_free(uio->uio_vio[ix].vio_base);
                }
                gsh_free(uio);
        }
```

we follow same in xdr_ioq_putbufs, where we increment in loop but decrement once.

Reply:
While currently, uio_count is always 1, it could become > 1 if we ever allow vector read/write. I'd rather make sure the code works OK if uio_count is ever > 1.

Reply:
I think we use xdr_ioq_uv for vector IO.
Also I added assert for assert((uio->uio_references == 1)) so that we won't proceed if uio_count > 1